// snyk/dist/cli/commands/test/iac-local-execution/local-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cleanLocalCache = exports.initLocalCache = exports.getLocalCachePath = exports.REQUIRED_LOCAL_CACHE_FILES = exports.LOCAL_POLICY_ENGINE_DIR = void 0;
const path = require("path");
const fs = require("fs");
const types_1 = require("./types");
const needle = require("needle");
const rimraf = require("rimraf");
const file_utils_1 = require("./file-utils");
const Debug = require("debug");
const debug = Debug('iac-local-cache');
exports.LOCAL_POLICY_ENGINE_DIR = '.iac-data';
const KUBERNETES_POLICY_ENGINE_WASM_PATH = path.join(exports.LOCAL_POLICY_ENGINE_DIR, 'k8s_policy.wasm');
const KUBERNETES_POLICY_ENGINE_DATA_PATH = path.join(exports.LOCAL_POLICY_ENGINE_DIR, 'k8s_data.json');
const TERRAFORM_POLICY_ENGINE_WASM_PATH = path.join(exports.LOCAL_POLICY_ENGINE_DIR, 'tf_policy.wasm');
const TERRAFORM_POLICY_ENGINE_DATA_PATH = path.join(exports.LOCAL_POLICY_ENGINE_DIR, 'tf_data.json');
exports.REQUIRED_LOCAL_CACHE_FILES = [
    KUBERNETES_POLICY_ENGINE_WASM_PATH,
    KUBERNETES_POLICY_ENGINE_DATA_PATH,
    TERRAFORM_POLICY_ENGINE_WASM_PATH,
    TERRAFORM_POLICY_ENGINE_DATA_PATH,
];
function doesLocalCacheExist() {
    return exports.REQUIRED_LOCAL_CACHE_FILES.every(fs.existsSync);
}
function getLocalCachePath(engineType) {
    switch (engineType) {
        case types_1.EngineType.Kubernetes:
            return [
                `${process.cwd()}/${KUBERNETES_POLICY_ENGINE_WASM_PATH}`,
                `${process.cwd()}/${KUBERNETES_POLICY_ENGINE_DATA_PATH}`,
            ];
        case types_1.EngineType.Terraform:
            return [
                `${process.cwd()}/${TERRAFORM_POLICY_ENGINE_WASM_PATH}`,
                `${process.cwd()}/${TERRAFORM_POLICY_ENGINE_DATA_PATH}`,
            ];
    }
}
exports.getLocalCachePath = getLocalCachePath;
async function initLocalCache() {
    const preSignedUrl = 'https://cloud-config-policy-bundles.s3-eu-west-1.amazonaws.com/bundle.tar.gz';
    file_utils_1.createIacDir();
    const response = needle.get(preSignedUrl);
    await file_utils_1.extractBundle(response);
    if (!doesLocalCacheExist()) {
        throw Error(`Missing IaC local cache data, please validate you have: \n${exports.REQUIRED_LOCAL_CACHE_FILES.join('\n')}`);
    }
}
exports.initLocalCache = initLocalCache;
function cleanLocalCache() {
    // path to delete is hardcoded for now
    const iacPath = path.join(`${process.cwd()}`, '.iac-data');
    try {
        // when we support Node version >= 12.10.0 , we can replace rimraf
        // with the native fs.rmdirSync(path, {recursive: true})
        rimraf.sync(iacPath);
    }
    catch (e) {
        debug('The local cache directory could not be deleted');
    }
}
exports.cleanLocalCache = cleanLocalCache;
